internal/controller/argocdcommitstatus_controller.go

```diff
@@ -0,0 +1,371 @@
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.19.1/pkg/reconcile
+func (r *ArgoCDCommitStatusReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling ArgoCDCommitStatus")
+
+	var argoCDCommitStatus promoterv1alpha1.ArgoCDCommitStatus
+	err := r.Get(ctx, req.NamespacedName, &argoCDCommitStatus, &client.GetOptions{})
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			logger.Info("ArgoCDCommitStatus not found")
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "failed to get ArgoCDCommitStatus")
+		return ctrl.Result{}, fmt.Errorf("failed to get ArgoCDCommitStatus: %w", err)
+	}
+
+	aggregates := map[objKey][]*Aggregate{}
+
+	ls, err := metav1.LabelSelectorAsSelector(argoCDCommitStatus.Spec.ApplicationSelector)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to parse label selector: %w", err)
+	}
+	// TODO: we should setup a field index and only list apps related to the currently reconciled app
+	var ul unstructured.UnstructuredList
+	ul.SetGroupVersionKind(gvk)
+	err = r.Client.List(ctx, &ul, &client.ListOptions{
+		LabelSelector: ls,
+	})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to list CommitStatus objects: %w", err)
+	}
+
+	ps, err := r.getPromotionStrategy(ctx, argoCDCommitStatus.GetNamespace(), argoCDCommitStatus.Spec.PromotionStrategyRef)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get PromotionStrategy %s: %w", argoCDCommitStatus.Spec.PromotionStrategyRef, err)
+	}
+
+	scmProvider, secret, err := utils.GetScmProviderAndSecretFromRepositoryReference(ctx, r.Client, ps.Spec.RepositoryReference, ps)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get ScmProvider and secret for ArgoCDCommitStatus %q: %w", argoCDCommitStatus.Name, err)
+	}
+
+	gitAuthProvider, err := r.getGitAuthProvider(ctx, scmProvider, secret)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get git auth provider for ScmProvider %q: %w", scmProvider.Name, err)
+	}
+
+	argoCDCommitStatus.Status.ApplicationsSelected = []promoterv1alpha1.SelectedApplications{}
+	for _, obj := range ul.Items {
+		var application argocd.ArgoCDApplication
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(obj.Object, &application)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to cast unstructured object to typed object: %w", err)
+		}
+
+		if application.Spec.SourceHydrator == nil {
+			continue
+		}
+
+		aggregateItem := &Aggregate{
+			application: &application,
+		}
+
+		key := objKey{
+			repo:         strings.TrimRight(application.Spec.SourceHydrator.DrySource.RepoURL, ".git"),
+			targetBranch: application.Spec.SourceHydrator.SyncSource.TargetBranch,
```

Comment:
Will eventually probably need URL normalization.

Reply:
Yea for sure

Reply:
Promotor always needs the https url pretty sure so we need to get it into that form from say `git@*` etc etc

Reply:
Yeah, I think Argo's normalization function always goes to https